#### Perform a Simple Linear Regression based on the Salary Dataset


##### Step 1: Import  all the necessary libraries

In [1245]:
import  numpy               as  np
import  pandas              as  pd
import  matplotlib.pyplot   as  plt
import  seaborn             as  sns

from    sklearn.model_selection     import  train_test_split
from    sklearn.preprocessing       import  StandardScaler
from    sklearn.linear_model        import  LinearRegression
from    sklearn.metrics             import  mean_absolute_error, mean_squared_error, root_mean_squared_error, r2_score

#### Step 2: Load the Dataset

In [1246]:
df = pd.read_csv("Salary_Data.csv")

In [1247]:
df

,YearsExperience,Salary
0,1.1,39343.0
1,1.3,46205.0
2,1.5,37731.0
3,2.0,43525.0
4,2.2,39891.0
5,2.9,56642.0
6,3.0,60150.0
7,3.2,54445.0
8,3.2,64445.0
9,3.7,57189.0


#### OBSERVATIONS:

1.  Here a dataset is provided for employees salary.

2.  The salaries for all the employees are provided based on the years of experience.

#### Step 3: Perform the Data Exploration

In [1248]:
#### get the first five rows of the dataset

print(df.head())

   YearsExperience   Salary
0              1.1  39343.0
1              1.3  46205.0
2              1.5  37731.0
3              2.0  43525.0
4              2.2  39891.0


In [1249]:
#### get the last five rows of the dataset

print(df.tail())

    YearsExperience    Salary
25              9.0  105582.0
26              9.5  116969.0
27              9.6  112635.0
28             10.3  122391.0
29             10.5  121872.0


In [1250]:
#### get the length of the dataset

print("Length of the dataset is:", len(df))

Length of the dataset is: 30


In [1251]:
#### get the shape of the dataset

print("Shape of the dataset is:", df.shape)

Shape of the dataset is: (30, 2)


In [1252]:
### get the information about the columns used in the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   YearsExperience  30 non-null     float64
 1   Salary           30 non-null     float64
dtypes: float64(2)
memory usage: 608.0 bytes


In [1253]:
#### get the descriptive summary statistics about the dataset

df.describe()

,YearsExperience,Salary
count,30.000000,30.000000
mean,5.313333,76003.000000
std,2.837888,27414.429785
min,1.100000,37731.000000
25%,3.200000,56720.750000
50%,4.700000,65237.000000
75%,7.700000,100544.750000
max,10.500000,122391.000000


In [1254]:
#### get the correlation between all the columns used in the dataset

df.corr()

,YearsExperience,Salary
YearsExperience,1.000000,0.978242
Salary,0.978242,1.000000


#### OBSERVATIONS:

1.  There is a very high correlation of 0.97 between the YearsExperience and Salary.

#### Step 4: Perform Data Cleaning

In [1255]:
#### Check for any NULL records in the dataset

df.isnull().sum()

YearsExperience    0
Salary             0
dtype: int64

#### OBSERVATIONS:

1.  There are no NULL records left in the dataset.

In [1256]:
#### Check for any duplicate records in the dataset

df[df.duplicated()]

,YearsExperience,Salary


#### OBSERVATIONS:

1.  There are no duplicate records left in the dataset.

#### Step 5: Divide the dataset into independent and dependent features

In [1257]:
##### Independent features

X = df.drop(columns='Salary', axis=1)

print(X)

    YearsExperience
0               1.1
1               1.3
2               1.5
3               2.0
4               2.2
5               2.9
6               3.0
7               3.2
8               3.2
9               3.7
10              3.9
11              4.0
12              4.0
13              4.1
14              4.5
15              4.9
16              5.1
17              5.3
18              5.9
19              6.0
20              6.8
21              7.1
22              7.9
23              8.2
24              8.7
25              9.0
26              9.5
27              9.6
28             10.3
29             10.5


In [1258]:
#### Dependent features

Y = df['Salary']

print(Y)

0      39343.0
1      46205.0
2      37731.0
3      43525.0
4      39891.0
5      56642.0
6      60150.0
7      54445.0
8      64445.0
9      57189.0
10     63218.0
11     55794.0
12     56957.0
13     57081.0
14     61111.0
15     67938.0
16     66029.0
17     83088.0
18     81363.0
19     93940.0
20     91738.0
21     98273.0
22    101302.0
23    113812.0
24    109431.0
25    105582.0
26    116969.0
27    112635.0
28    122391.0
29    121872.0
Name: Salary, dtype: float64


#### Step 6: Divide the independent and dependent variables into training and testing data

In [1259]:
from  sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2,random_state=42)

In [1260]:
X_train

,YearsExperience
28,10.3
24,8.7
12,4.0
0,1.1
4,2.2
16,5.1
5,2.9
13,4.1
11,4.0
22,7.9


In [1261]:
X_test

,YearsExperience
27,9.6
15,4.9
23,8.2
17,5.3
8,3.2
9,3.7


In [1262]:
print("Shape of the input training data is:", X_train.shape)

print("Shape of the input testing data is:",  X_test.shape)

Shape of the input training data is: (24, 1)
Shape of the input testing data is: (6, 1)


In [1263]:
Y_train

28    122391.0
24    109431.0
12     56957.0
0      39343.0
4      39891.0
16     66029.0
5      56642.0
13     57081.0
11     55794.0
22    101302.0
1      46205.0
2      37731.0
25    105582.0
3      43525.0
21     98273.0
26    116969.0
18     81363.0
29    121872.0
20     91738.0
7      54445.0
10     63218.0
14     61111.0
19     93940.0
6      60150.0
Name: Salary, dtype: float64

In [1264]:
Y_test

27    112635.0
15     67938.0
23    113812.0
17     83088.0
8      64445.0
9      57189.0
Name: Salary, dtype: float64

In [1265]:
print("Shape of the output training data is:", Y_train.shape)

print("Shape of the output testing data is:",  Y_test.shape)

Shape of the output training data is: (24,)
Shape of the output testing data is: (6,)


#### Step 7: Perform Feature Scaling on the inputs

In [1266]:
from   sklearn.preprocessing    import  StandardScaler

#### create an object for  Standard Scaler

sc = StandardScaler()


X_train_scaled = sc.fit_transform(X_train)


X_test_scaled  = sc.transform(X_test)

In [1267]:
X_train_scaled

array([[ 1.77445774],
       [ 1.21912622],
       [-0.41216011],
       [-1.41869849],
       [-1.03690807],
       [-0.03036969],
       [-0.79395053],
       [-0.37745189],
       [-0.41216011],
       [ 0.94146047],
       [-1.34928205],
       [-1.27986561],
       [ 1.32325088],
       [-1.10632451],
       [ 0.66379471],
       [ 1.49679198],
       [ 0.24729607],
       [ 1.84387418],
       [ 0.55967005],
       [-0.68982587],
       [-0.44686833],
       [-0.23861901],
       [ 0.28200429],
       [-0.75924231]])

In [1268]:
X_test_scaled

array([[ 1.5315002 ],
       [-0.09978613],
       [ 1.04558513],
       [ 0.03904675],
       [-0.68982587],
       [-0.51628477]])

#### OBSERVATIONS:

1.  The standard scaling has been applied on all the inputs and all these inputs has been converted in the same scale range.

#### Step 8: Train the model using Linear Regression

In [1269]:
#### create an object for Linear Regression

lin = LinearRegression()


#### using the object for Linear Regression, train the model

lin.fit(X_train_scaled, Y_train)

LinearRegression()

#### Step 9: Predict the Model

In [1270]:
Y_pred = lin.predict(X_test_scaled)

In [1271]:
Y_pred

array([115790.21011287,  71498.27809463, 102596.86866063,  75267.80422384,
        55477.79204548,  60189.69970699])

#### Step 10: Evaluate the Regression Metrics

In [1272]:
from    sklearn.metrics             import  mean_absolute_error, mean_squared_error, root_mean_squared_error, r2_score


mae = mean_absolute_error(Y_test, Y_pred)

print("Mean Absolute Error of the model is:", mae)

Mean Absolute Error of the model is: 6286.453830757746


In [1273]:
mse = mean_squared_error(Y_test, Y_pred)

print("Mean Squared Error of the model is:", mse)

Mean Squared Error of the model is: 49830096.85590834


In [1274]:
rmse = root_mean_squared_error(Y_test, Y_pred)

print("Root Mean Squared Error of the model is:", rmse)

Root Mean Squared Error of the model is: 7059.043621901506


In [1275]:
r2 = r2_score(Y_test, Y_pred)

print("R2 Score of the model is:", r2)

R2 Score of the model is: 0.9024461774180498


#### OBSERVATIONS:

1.  The R2 Score of the model using Linear Regression is 0.9024.  It is very high. So the model is working completely fine.

#### Step 11: Apply Ridge Regression 

#### (a.) Ridge Regression 

Ridge Regression is used to prevent overfitting

In [1276]:
from  sklearn.linear_model  import Ridge


#### create an object for Ridge regression

ridge = Ridge(alpha=1.0)


#### using the object for ridge , train the model

ridge.fit(X_train_scaled, Y_train)

Ridge()

In [1277]:
#### Predict the model

Y_pred_ridge = ridge.predict(X_test_scaled)

print(Y_pred_ridge)

[114126.90670836  71606.65197084 101461.29891421  75225.39705489
  56226.98536366  60750.41671871]


In [1278]:
#### Evaluate the regression metrics

from    sklearn.metrics             import  mean_absolute_error, mean_squared_error, root_mean_squared_error, r2_score


mae_ridge = mean_absolute_error(Y_test, Y_pred_ridge)

print("Mean Absolute Error of the model is:", mae_ridge)

Mean Absolute Error of the model is: 6192.215677527437


In [1279]:
mse_ridge = mean_squared_error(Y_test, Y_pred_ridge)

print("Mean Squared Error of the model is:", mse_ridge)

Mean Squared Error of the model is: 51710764.81670462


In [1280]:
rmse_ridge = root_mean_squared_error(Y_test, Y_pred_ridge)

print("Root Mean Squared Error of the model is:", rmse_ridge)

Root Mean Squared Error of the model is: 7191.019734134


In [1281]:
r2_ridge = r2_score(Y_test, Y_pred_ridge)

print("R2 Score  of the model is:", r2_ridge)

R2 Score  of the model is: 0.8987643393290411


#### OBSERVATIONS:

1.  The R2 Score of the model using Ridge Regression is 0.8987 .  It is very high. So the model is working completely fine.

##### (b.) Lasso Regression 

Lasso Regression is used to perform feature selection.

In [1282]:
from  sklearn.linear_model  import Lasso


#### create an object for Ridge regression

lasso = Lasso(alpha=0.1)


#### using the object for ridge , train the model

lasso.fit(X_train_scaled, Y_train)

Lasso(alpha=0.1)

In [1283]:
#### Predict the model

Y_pred_lasso = lasso.predict(X_test_scaled)

print(Y_pred_lasso)

[115790.05696285  71498.28807324 102596.76410212  75267.80031917
  55477.86102806  60189.75133547]


In [1284]:
#### Evaluate the regression metrics

from    sklearn.metrics             import  mean_absolute_error, mean_squared_error, root_mean_squared_error, r2_score


mae_lasso = mean_absolute_error(Y_test, Y_pred_lasso)

print("Mean Absolute Error of the model is:", mae_lasso)

Mean Absolute Error of the model is: 6286.445153702766


In [1285]:

mse_lasso = mean_squared_error(Y_test, Y_pred_lasso)

print("Mean Squared Error of the model is:", mse_lasso)

Mean Squared Error of the model is: 49830194.13598571


In [1286]:

rmse_lasso = root_mean_squared_error(Y_test, Y_pred_lasso)

print("Root Mean Squared Error of the model is:", rmse_lasso)

Root Mean Squared Error of the model is: 7059.050512355448


In [1287]:
r2_lasso = r2_score(Y_test, Y_pred_lasso)

print("R2 Score  of the model is:", r2_lasso)

R2 Score  of the model is: 0.9024459869700273


#### OBSERVATIONS:

1.  The R2 Score of the model using Lasso Regression is 0.9024 .  It is very high. So the model is working completely fine.

##### (c.)  Elastic Net Regression 


Elastic Net Regression is done

   (a.)   To prevent the overfitting

   (b.)   To perform feature selection

   (c.)   To perform hyper parameter tuning of the model


In [1288]:
from  sklearn.linear_model  import ElasticNet


#### create an object for Ridge regression

elasticnet = ElasticNet(alpha=0.1, l1_ratio=0.5)


#### using the object for ridge , train the model

elasticnet.fit(X_train_scaled, Y_train)

ElasticNet(alpha=0.1)

In [1289]:
#### Predict the model

Y_pred_elasticnet = elasticnet.predict(X_test_scaled)

print(Y_pred_elasticnet)

[113810.01408368  71627.29936565 101244.95012512  75217.31763953
  56369.72170169  60857.24454403]


In [1290]:
#### Evaluate the regression metrics

from    sklearn.metrics             import  mean_absolute_error, mean_squared_error, root_mean_squared_error, r2_score


mae_elasticnet = mean_absolute_error(Y_test, Y_pred_elasticnet)

print("Mean Absolute Error of the model is:", mae_elasticnet)

Mean Absolute Error of the model is: 6174.261421171672


In [1291]:
mse_elasticnet = mean_squared_error(Y_test, Y_pred_elasticnet)

print("Mean Absolute Error of the model is:", mse_elasticnet)

Mean Absolute Error of the model is: 52256018.15224882


In [1292]:
rmse_elasticnet = root_mean_squared_error(Y_test, Y_pred_elasticnet)

print("Root Mean Absolute Error of the model is:", rmse_elasticnet)

Root Mean Absolute Error of the model is: 7228.8324197098955


In [1293]:
r2_elasticnet = r2_score(Y_test, Y_pred_elasticnet)

print("R2 Score of the model is:", r2_elasticnet)

R2 Score of the model is: 0.8976968810956825


#### OBSERVATIONS:

1.  The R2 Score of the model using Elastic Net Regression is 0.8976 .  It is very high. So the model is working completely fine.


2.   The R2 Score of the Linear Regression = Lasso Regression  > R2 Score of Ridge = Elastic Regression


3.   (a.)   The R2 Score of the Linear       Regression = 0.9024

     (b.)   The R2 Score of the Lasso        Regression = 0.9024

     (c.)   The R2 Score of the Ridge        Regression = 0.8976
     
     (d.)   The R2 Score of the Elastic Net  Regression = 0.8976